### DISCUSSION TITLE
Pico 2/RP2350 Support?

### DISCUSSION THREAD

**Discussion by phil-barrett** [grblHAL/core#567] [General]
This is a nice evolution of the RP2040. Adds FPU and more PIOs. I'd be happy to help testing and update PicoCNC to support it.

**Comment by JosephRBiden** [comment]
I agree, but It would especially be nice to have support for the RP2350B variant, as this carries and extra 18 GPIO pins compared to the A model, and 2040, which would allow for more accessories, use in more machines, and/or allowing for a good 4 axis board, without using shift registers, which could allow for an even cheaper and smaller option.

**Comment by terjeio** [comment]
> I'd be happy to help testing and update PicoCNC to support it.

The pinout of Pi Pico 2 is supposedly the same as the Pi Pico so my guess is that initially it will be a matter of compiling for it, perhaps with tweaking a few clocks to get timings right. 

> ... It would especially be nice to have support for the RP2350B variant

If somebody comes up with board design(s) I'll add the code...

**Reply by phil-barrett** [reply]
Looking into the 2350 as we speak.

**Reply by phil-barrett** [reply]
Pico 2 doesn't seem to be available in the US yet.  I have a couple on pre-order.  

The RP2350B (80 pin package) has 48 GPIOs so we could skip the SRs and support 5 Axes with limits, I2C (inc strobe), SPI, 4 Aux in, 4 Aux out, 2 probes, Rx/Tx, ethernet, standard PWM spindle with a few pins left over. With the FPU, we could support spindle sync.

Now we get to the reality point. None of the RP2350 chips appear to be available yet - A (QFN60), B (QFN80) and RP2354 A or B (with 2MB flash) don't show up in the usual distis.  Various sources say end of 2024 for all the chips but I suspect the 2350s will be sooner than that. I can't believe that the Pico 2 is delayed that long.

**Reply by JosephRBiden** [reply]
<img width="437" alt="Screen Shot 2024-08-10 at 11 20 47 AM" src="https://github.com/user-attachments/assets/109d7d02-1dbe-4cb0-90a9-ad80217fb8e8">

**Reply by JosephRBiden** [reply]
I asked them if the 2350Bs would be available, this is the response I got. Not very sure what the sample program is, but the form is the thing on the bottom of the page about the 2350s, or when you click "register interest" on their page...

**Comment by phil-barrett** [comment]
A bit more digging this morning.  There are a number of products being sold (at least advertised) with the RP2350B on them. In particular, there is the [PGA2350](https://shop.pimoroni.com/products/pga2350?variant=42092629229651) carrier board that has the RP2350B on it.  The image of the chip says it is an A2 stepping which implies some level of stability. I think the outfit (Pimoroni, UK) is fairly reputable so I bought a couple with DHL shipping.  If they are really in stock, I should have them in my hand this coming week. Am laying out a test board for it.  Won't sell this board as I don't trust the PGA2350 to have long availability but worth it to get some testing in before the RP2350B becomes widely available (especially to my CM in China).

![image](https://github.com/user-attachments/assets/c9fa8a7d-deb0-4a39-9656-0d47e587d6cc)

![image](https://github.com/user-attachments/assets/614ff096-f1b9-4b6e-b2cc-df9a5f2212ec)

**Reply by JosephRBiden** [reply]
Yea, I saw the same thing. I bought 4 of those on Friday, but they haven't shipped yet, I guess they don't ship out on weekends...

**Comment by phil-barrett** [comment]
Here's an initial pin mapping for the RP2350B.  Note that it uses KiCAD global labels in an odd way.  The purpose of that is to allow quick reorganization during design and make for an easy way to move from the PGA2350 board to the actual RP2350B when available for board manufacturing. Pin assignment has been done by groups to make general routing easier

I wasn't able to get to 4 aux input pins but think 3 will be sufficient.  Not sure how I will condition them.  The RP2350 GPIOs are 5V tolerant when the chip is running at 3.3V so maybe just an LPF? They appear to be Schmitt triggered.

Also, I wonder how useful individual Enables are for each axis.  Would a single Enable for all be acceptable?  I could gain back 4 pins. 

![image](https://github.com/user-attachments/assets/b721e9a8-492b-4940-b5a8-88cf16394033)

**Reply by terjeio** [reply]
> Also, I wonder how useful individual Enables are for each axis. Would a single Enable for all be acceptable? 

IMO at least keep the Z enable separate since some spindles get pulled down by gravity if the stepper is disabled.

And please keep the step pins consecutive for [simpler and more compact PIO code](https://github.com/grblHAL/RP2040/blob/343a2e3b0dc18c44761ad0a0d3edcf14aec3e7bc/driverPIO.pio#L70-L108) - only one state machine is needed if you do so. If not one is required for each axis.

**Reply by phil-barrett** [reply]
PIO is used even if no shift regs? 

I've kept separate enables. But, is there a way to have assignable enables? Maybe have 2 enable pins that can be assigned to different axes?

**Reply by terjeio** [reply]
> PIO is used even if no shift regs?

Yes, since it reduces interrupt overhead - normally two (or three if delayed pulse start) interrupts are fired per step pulse, only one when PIO is used since the PIO code handles the timing and writing to the pins. I _can_ add support for the "standard" way of outputting pulses if/when the state machines will be needed for other purposes.

> But, is there a way to have assignable enables? 

Currently not, other than one for Z and one for the rest. I have to check the RP2040 driver to see if it supports this - if not I'll add it since some of the other drivers do.

**Reply by phil-barrett** [reply]
No problem to make all the step pins consecutive. 

I've kept 5 enables since I'm not sure what I would use the extra pins for. Maybe additional SPI CS pins? Or more Digital I/O (I have 6  - 3 in, 3 out)? I'm just not sure that would make much of a difference.

I'm about halfway done with the first cut of the design.  The Pimoroni carrier board makes a decent jump start. It pretty closely follows the RP2350B pinout so it should be relatively easy to drop the chip in place of it. Will do a quick layout with the chip to make sure everything will fit. I'm keeping the board under 10x10 cm - that seems to be a magic number in more ways than one.

I'm thinking of getting rid of a number of the pin headers - step signals and spindle group. Very few of my customer use them and they take up space. Feedback appreciated.

**Reply by terjeio** [reply]
> I've kept 5 enables since I'm not sure what I would use the extra pins for. Maybe additional SPI CS pins?

Add 3 digital inputs that can be used for an encoder (A/B/index or select)? For spindle sync or the encoder plugin. Alternatively a separate SPI port? A performance booster for networking would be if the SD card and the W5500 were on separate busses.
At least one input that can be configured for analog in?

> Will do a quick layout with the chip to make sure everything will fit.

You are aware of the recommended/required layout for the regulator components? Section 6.3.8 in the datasheet has the details.

**Reply by phil-barrett** [reply]
Yes, I've been looking at the VReg issues. The HW design guide has a section on it as well. There is a fair amount of consternation out there over the issue of the inductor orientation and only a single supplier. eevblog has a long thread on it with some saying the chip is too broken to use. Plus, almost no inductors for this kind of application are marked for polarity and HW design guide says that most parts aren't even loaded onto tape in a consistent orientation. There isn't even a part number yet for the Abracom part that they recommend. Abracom parts tend to be super expensive in China - almost as much as the RP chip itself. But, I'm sure it will get sorted out in due order.

The current design has 3 digital inputs.  I'll take a look at a separate SPI bus though that's 4 pins right there.

**Reply by andrewmarles** [reply]
If power consumption isn't critical (like on a CNC) you can just use an external LDO for the 1.1V.

**Reply by phil-barrett** [reply]
> If power consumption isn't critical (like on a CNC) you can just use an external LDO for the 1.1V.

Yeah, the more I look at this I don't understand the designers' reasoning.  For hobby and small system use, I don't see why they need a SMPS for 1.1V.  There are 1.1V peripherals, maybe they expect more use of those?  The 2350 only provides switcher logic and that saves a pretty small amount of board space.

**Reply by andrewmarles** [reply]
The 1.1V is for the core.  The SMPS is to reduce power consumption of their device, which is one of the biggest things that keeps the RP2040 from being used in real commercial products.  You could of course implement an external 1.1V SMPS, but at that point you might as well use a chip from another vendor that doesn't have this problem.

For hobbyists and powered devices it is cheaper and simpler to just use an LDO; scanning quickly there are options that cost about 3-5 cents.

Personally I was hoping for more pins.  But aside from that, why not use shift registers?  This is a strength of the PIO, not something to avoid.  Has anyone ever given @Volksolive's 6 axis shift register PIO code a try?

https://github.com/Volksolive/RP2040/blob/43a232ce501498f22810a632643f56732967355a/driverPIO.pio#L111

**Reply by phil-barrett** [reply]
I haven't used it but my gut instinct says that for 2 SRs the performance will fall off.  Right now, with 4 axes, I'm seeing about 230 kHz. To me, 200 kHz is kind of the lower bound for a good motion controller.

**Reply by phil-barrett** [reply]
By the way, the data sheet does not show how to use an LDO for DVDD in place of the SMPS. I'm guessing that you don't connect VREG_FB and VREG_LX.  Not clear about VREG_PGND, probably connect to ground to prevent floating circuitry.  I assume you do connect VREG_IN and VREG_AVDD to 3.3V (assuming that's what IOVDD is using).

**Reply by andrewmarles** [reply]
You can find a pretty detailed example of driving the 1.1V with an LDO here:

https://hardware.buspirate.com/rp2350

**Reply by phil-barrett** [reply]
Thanks. Their blog goes into some detail.  Looks like VREG_AVDD and VREG_VIN are fed filtered 3.3V, the other 3 VREG pins are left unconnected (including VREG_PGND). The blog mentions that, initially, they didn't connect VREG_IN (and the board worked) but the RP people later said to connect it.  The HW design guide really should have that info after their "scare tactics" on the inductor choice.

**Reply by phil-barrett** [reply]
By the way the Bus Pirate guys are using the 2350B on a version of the BP (BP 6). In the blog the designer goes on about how satisfying it was to get rid of their shift regs. Space was a bit tight for them. I chuckled over that because of the SR discussion we had.

**Comment by phil-barrett** [comment]
Quick update on the 2350B board.  Design for the PGA2350 is done - it fits on a 2 layer board though I might move to 4 layer for cleaner power distribution. The current design uses a buck converter for 5V which I am not happy with.  I'll probably use a buck for 3.3V and an LDO for 5V and 1.1V. There are a lot of potential users of 3.3V so I want that to be pretty robust.  I have a revised pin out shown below.  I may tweak it a bit going forward (moving the SPI and I2C pins around a bit to make for a simpler layout.

![image](https://github.com/user-attachments/assets/ed2f1436-92b8-470e-a49e-6f01dbc9e066)

Here's the current board with the Ethernet adaptor and an (admittedly crude) PGA2350 model.
![image](https://github.com/user-attachments/assets/6d604655-3ebc-42ba-bfac-9581ffa92fe3)

**Reply by terjeio** [reply]
Can aux in 0/1 be optionally used as a second UART port? For those who have a RS485 VFD that want to use a MPG.

**Reply by phil-barrett** [reply]
Not with this pin assignment.  Aux 0/1 can be used for UART 0 but the the other UART pins are for UART 0 as well.  

However, I think this is a good idea and worth doing.  I might be able to swap some pins around.  I'll need to make some changes to the HW to allow output (if I use aux in) or input (if aux out) on one of the pins.

**Comment by phil-barrett** [comment]
Update on progress.  

RP2350s are still not available in China so I'm not in a big hurry. I'm still hearing end of the year. But, I've got a prototype board back based on the Pimoroni PGA2350 and it seems to work well with initial testing.  0-10V is rock solid linear. Other than one bodge wire (stupid mistake) it's pretty clean. No grblHAL for it yet, though. I will not be releasing this board, will move to the RP2350B chip.
![PXL_20240922_183401941](https://github.com/user-attachments/assets/46edf49d-935a-4d2e-ac21-4ff7a7e58383)

It is very similar to the Teensy 4.1 board. It does not use shift registers.
Some quick specs:
- 5 axis - step dir, enable, limit
- 7 relay coil support with 5/12V selection
- 5V outputs for Flood, Mist, AuxOut, Spindle Enable, Vac/DC
- 3 digital outputs (AuxOut0-AuxOut2)
- 3 digital inputs (AuxIn0-AuxIn2) with LPF. 5V tolerant.
- Standard Grlb control inputs - Start, Hold, Estop/Halt, Door, Probe
- 2 UARTs, UART0 and UART1.  UART1 Steals AuxIn2 and AuxOut2 via board level config.
- UART0 selectable for 3.3V or 5V.
- I2C Header with strobe and Halt pass through.
- SPI Header
- WizNet Ethernet Support
- Spindle support - Enable, Direction, 0-10V, PWM, experimenting with open collector for EN and Dir

I have removed some pin headers.  It appears very few people are using them for Stepper Driver and VFD interface.

I am hoping that grblHAL on this board will support spindle sync to make it a good lathe board.

Pin Map:
![pin assignments](https://github.com/user-attachments/assets/e9051f61-943f-4235-a5c2-4965962bb026)

**Comment by axxel** [comment]
I'm about to build a standard 3-axis table CNC machine and am looking for the best "driver" hardware option. It seems to me the most mature and reliable option would be the ESP32? But since I'm using the RP2040 for another project with my own firmware already, I'd prefer that one to spare me the dive into yet another SDK and platform in case I want to contribute down the road. Based on the "stars" in github, RP2040 seems actually more popular, too.

Questions:
1. is there a clear advantage of one over the other?
2. does the arrival of the RP2350 on the scene change that picture significantly, considering the FPU unit and the higher clock speed?
3. what is the current status of the RP2350 support?

(Disclaimer: If this is considered off topic in this discussion, I'll start a new one)

**Reply by phil-barrett** [reply]
Probably better to start a separate discussion as esp32  fans would not see it here.

But, I will answer your questions here.

RP2040 and Pico is quite mature and reliable at this point. ESP maturity depends on the version but generally have a similar level of maturity and reliability.

RP2350 is fairly new and grblHAL support is also new.  The Web Builder now has support for the Pico2 on the PicoCNC board. I have tested it and it works fairly well.  I still have some performance testing to do with the build but expect it will be somewhat faster than the Pico (RP2040) version but not dramatically so..  The good news is you can use either the Pico (RP2040) or Pico2 (RP2350) with the PicoCNC board.  I am hoping to eventually release an RP2450 base 5 Axis board.  It is at minimum a few months away.

I don't believe the Pico2 (RP2350) will significantly speed up grblHAL.  I believe the Pico (RP2040) is on a par with the ESP32 in terms of performance.  The RP2350 based 5 Axis board looks like it will have significantly higher step rate performance, FWIW.

**Reply by axxel** [reply]
Thanks for your insights. There are two claims that I find surprising though:
1) RP2040 on par with ESP32 in terms of performance, even though the latter has twice the clock speed and an FPU?
2) Why would the RP2350 based 5 axis board provide a significantly higher step rate than the PicoCNC with a Pico2? Because of the PIO having to spend at least 8x the amount of cycles programming the SR compared to the direct GPIO approach?
3) Are there "official" max step rates to be found somewhere for the ESP, RP2040 and RP2350?

**Reply by phil-barrett** [reply]
From memory.  I saw general benchmarks somewhere. Sorry don't have a link.

The RP ports of grblHAL uses PIO for step generation which offloads a lot of work from the CPU.  The RP2350 PIO is faster than the RP2040. The 2350 has an FPU.  The RP2350 5 Axis board does not use shift regs. The RP2040 port for the PicoCNC board uses PIO for the shift reg.

There are no official benchmarks for step rate, btw. I've measure Teensy and PicoCNC. You are free to give it a try for other processors/boards.

**Reply by terjeio** [reply]
> is there a clear advantage of one over the other?

For the end user likely no.
For a developer yes, for now SDK updates from Raspberry Pi is has been backwards compatible, updates from Espressif not necessarily so sometimes causing a lot of trouble. Debug support for the Pico is excellent, for the ESP32 not so. Espressif also forces one to use FreeRTOS which, IMO, adds overhead and complexity.

> does the arrival of the RP2350 on the scene change that picture significantly, considering the FPU unit and the higher clock speed?

I think so, not only by having a FPU but also more PIOs, and the B variant having significantly more GPIO pins.
In practical terms this means the RP2350/54 can be more easily extended with optional features such as ethernet, spindle sync, step injection (for plasma THC and stepper driven spindle), general I/O (digital and analog) etc. The FPU also allows for more headroom to tune for faster laser raster engraving where lots of small movements are commanded.

> what is the current status of the RP2350 support?

Nearly complete compared to the RP2040 since the driver codebase has been updated to support[ both the RP2340 and the RP2350/54](https://github.com/search?q=repo%3AgrblHAL%2FRP2040+%22RP_MCU+%3D%3D+2040%22&type=code). Only RTC (real time clock) support is missing for the latter for now, causing files uploaded to a SD card to not be time stamped. And currently step pulse timing is a bit off for the RP2350/54, I have a pending commit that will fix that and add some other improvements for both.

> Are there "official" max step rates to be found somewhere for the ESP, RP2040 and RP2350?

These are somewhat hard to determine since many factors affect them: interrupt latency, FPU availablility/speed, planner buffer size, number of axes enabled and moving, ... There is a hard lower limit though, mainly determined by interrupt latency and how long it takes to output steps from the steps (segment) buffer. Using RP PIO or ESP32 RMT for step generation removes one interrupt latency/interrupt processing time from the equation allowing for higher step rates compared to what is possible with other processors/configurations.

**Reply by phil-barrett** [reply]
RP235x update.  I have received a batch of beta build V0.92 RP2350B boards.  So far in my testing, it works pretty well.  One thing to change is the LDO for 5V from 12V.  With the ethernet module, it gets a bit too hot for my liking.  Have replaced that with a buck converter.  For more detail, check out my [RP23CNC github repo](https://github.com/phil-barrett/RP23CNC/tree/main) for it.  I will be sending boards to beta testers in the next few days.  If you are interested in beta testing the board, contact me on the discussions section of the RP23CNC github repo.  I have a small number of boards available.

![T2120387_DxO](https://github.com/user-attachments/assets/c2b83de2-bf0f-474f-9271-8dac1c89d13a)

**Reply by axxel** [reply]
Thanks for background info! After lots of delays my machine is now finally actually moving. I went with a Pico/RP2040 I still had and basically a bread board. Very simple setup. I still have not come around to touch the code, yet. But just stumbled upon the news about the RP2040 now supporting [200MHz](https://github.com/raspberrypi/pico-sdk/releases/tag/2.1.1) via a simple `#define SYS_CLK_MHZ=200`. I was wondering whether any of you have given this a shot, yet?

**Reply by terjeio** [reply]
>  I was wondering whether any of you have given this a shot, yet?

I just tried and gave up for now - VSCode will not create the necessary build files when switching board from Pico2 W to Pico W or Pico so building does not even start.

**Reply by terjeio** [reply]
Got it working, a small typo was the cause - and the error I got was completely unrelated.
I had to change some code to get step timing correct, this will be in the next commit along with max frequency (200 MHz) automatically selected if building with SDK >= 2.1.1.

By changing the min. step pulse allowed temporarily to 1 microsecond I managed to push the max step rate to 400KHz for single axis straight moves:

![image](https://github.com/user-attachments/assets/c542ed4d-ccab-44f0-bbdf-d31547211e5b)

**Comment by phil-barrett** [comment]
That was with an RP2040 at 200 MHz?  Wow.

This makes me bring up a previous discussion point.  Should the overall grblHAL 2 uS pulse limit be lowered?  And, should the default of 10 uS also be lowered? I know the previous response has been to avoid making it possible for people to screw themselves.  However, I think you can't really idiot proof this because idiots can be quite determined.  I propose we have a 1 uS lower limit and a 2 uS default.  Add a stern warning about lower than 2 uS. Even the cheapest drivers I am aware of handle 2 uS just fine.

**Reply by terjeio** [reply]
>  Should the overall grblHAL 2 uS pulse limit be lowered?

If so it has to be driver specific because most cannot go much below 2 uS due to interrupt latency. The RP2040/2350 and ESP32 (with RMT generated step pulses) are the only ones that does not use a second interrupt for turning off the step pulses - it is done without processor intervention.

>  And, should the default of 10 uS also be lowered?

This should be less problematic.

**Reply by phil-barrett** [reply]
Fair point though doesn't the iMXRT1062 also allow shorter pulses?  I recall someone (Jim Fong?) testing a custom build down to about 1 uS.

> > And, should the default of 10 uS also be lowered?

> 
> This should be less problematic.

Definitely the more beneficial one.  10 uS limits to around 80 kHz.

**Reply by axxel** [reply]
Interesting subject. I was curious what my DM542 have to say. I actually could not find proper information about this from the vendor I got them from (ACT Motors) but the 'original' (Leadshine) datasheet says 2.5µs minimal pulse length at max 200kHZ.

If I understand correctly, the default can currently not be changed via the webbuilder, right?

**Reply by phil-barrett** [reply]
It's a Grbl Setting - $0.  Default is set in the code. 

I'm pretty sure those are ultra conservative specs - most can handle much faster.

**Reply by terjeio** [reply]
> Fair point though doesn't the iMXRT1062 also allow shorter pulses?

True, it is the only one I have tested, some years ago now, that had sub microsecond interrupt latency.

**Comment by phil-barrett** [comment]
By the way, I am moving my RP235x discussions to the [RP2040 driver repo](https://github.com/grblHAL/RP2040/discussions).  Probably doesn't belong in grblHAL/Core.

I have updates on the RP2350B beta boards there.
